File: loop_rooms/loop_rooms.py
import sys
sys.setrecursionlimit(10000) #too much, but we think enough for the larger test cases than grader :)
                             #for "maze17.txt:, sys.setrecursionlimit(3500) is OK

with open(sys.argv[1]) as f:
    lines = f.read().splitlines()

N,M = map(int, lines[0].split())
maze = list(map(list, lines[1:]))
counter=0;

for i in range (0,N,1):
    for j in range (0,M,1):
        if(j==0 and maze[i][j]=='L') or (j == M-1 and maze[i][j] == 'R') or (i == 0 and maze[i][j] == 'U') or (i == N-1 and maze[i][j] == 'D'):
            maze[i][j] = 'G'
            #counter=counter+1

n,m = N-1, M-1

def prev_room(maz, n, m, k, l):
    if(k==0 and l==0):                  #Upper-Left
        if maz[k][l+1]== 'L':
            maz[k][l+1]= 'G'
            prev_room(maz, n, m, k, l+1)
        if maz[k+1][l]=='U':
            maz[k+1][l]='G'
            prev_room(maz, n, m, k+1, l)
    elif k==0 and l==m :                #Upper-Right
        if maz[k][l-1]=='R':
            maz[k][l-1]='G'
            prev_room(maz, n, m, k, l-1)
        if maz[k+1][l]=='U':
            maz[k+1][l]='G';
            prev_room(maz, n, m, k+1, l)
    elif k==n and l==0 :                 #Bottom-Left
        if maz[k][l+1]=='L':
            maz[k][l+1]='G'
            prev_room(maz, n, m, k, l+1)
        if maz[k-1][l]=='D':
            maz[k-1][l]='G'
            prev_room(maz, n, m, k-1, l)
    elif k==n and l==m:                 #Bottom-Right
        if maz[k][l-1]=='R':
            maz[k][l-1]='G'
            prev_room(maz, n, m, k, l-1)
        if maz[k-1][l]=='D':
            maz[k-1][l]='G'
            prev_room(maz, n, m, k-1, l)
    elif k!=0 and k!=n and l==0:        #Left Side
        if maz[k][l+1]=='L':
            maz[k][l+1]='G'
            prev_room(maz, n, m, k, l+1)
        if maz[k-1][l]=='D':
            maz[k-1][l]='G'
            prev_room(maz, n, m, k-1, l)
        if maz[k+1][l]=='U':
            maz[k+1][l]='G'
            prev_room(maz, n, m, k+1, l)
    elif l==m and k!=0 and k!=n:        #Right Side
        if maz[k][l-1]=='R':
            maz[k][l-1]='G'
            prev_room(maz, n, m, k, l-1)
        if maz[k-1][l]=='D':
            maz[k-1][l]='G'
            prev_room(maz, n, m, k-1, l)
        if maz[k+1][l]=='U':
            maz[k+1][l]='G'
            prev_room(maz, n, m, k+1, l)
    elif k==0 and l!=0 and l!=m:        #Upper Side
        if maz[k][l-1]=='R':
            maz[k][l-1]='G'
            prev_room(maz, n, m, k, l-1)
        if maz[k][l+1]=='L':
            maz[k][l+1]='G'
            prev_room(maz, n, m, k, l+1)
        if maz[k+1][l]=='U':
            maz[k+1][l]='G'
            prev_room(maz, n, m, k+1, l)
    elif k==n and l!=0 and l!=m:        #Bottom Side
        if maz[k][l-1]=='R':
            maz[k][l-1]='G'
            prev_room(maz, n, m, k, l-1)
        if maz[k][l+1]=='L':
            maz[k][l+1]='G'
            prev_room(maz, n, m, k, l+1)
        if maz[k-1][l]=='D':
            maz[k-1][l]='G'
            prev_room(maz, n, m, k-1, l)
    else :                              #Inside
        if maz[k][l-1]=='R':
            maz[k][l-1]='G'
            prev_room(maz, n, m, k, l-1)
        if maz[k][l+1]=='L':
            maz[k][l+1]='G'
            prev_room(maz, n, m, k, l+1)
        if maz[k-1][l]=='D':
            maz[k-1][l]='G'
            prev_room(maz, n, m, k-1, l)
        if maz[k+1][l]=='U':
            maz[k+1][l]='G'
            prev_room(maz, n, m, k+1, l)





for i in range (0,N,1):
    for j in range (0,M,1):
        if ((j==0 or j == M-1 or i == 0 or i == N-1) and maze[i][j] == 'G') :
            prev_room(maze,N-1,M-1,i,j)

cnt=0

for i in range (0,N,1):
    for j in range (0,M,1):
        if maze[i][j]=='G':
            cnt=cnt+1
res=N*M-cnt
print(res)

File: loop_rooms/loop_rooms.cpp
#include <iostream>
#include <fstream>
using namespace std;

int prev_room(char maz[], bool fl[], int n, int m, int k, int l, int &cnt){
  /*recursive function:we find the previous room of a good room. Previous with the given directions.
  "Good" is the room from which you can go outside*/
  int rows;
  rows=m+1;
  if((k==0) && (l==0)){  //Upper-Left
    //k=i;
    //l=j;
    if((maz[k*rows+(l+1)]=='L')&&(fl[k*rows+(l+1)]==0)){
      fl[k*rows+(l+1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l+1, cnt);
    }
    if((maz[(k+1)*rows+l]=='U')&&(fl[(k+1)*rows+l]==0)){
      fl[(k+1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k+1, l,cnt);
    }
  }
  else if((k==0)&&(l==m)){  //Upper-Right
    //k=i;
    //l=j;
    if((maz[k*rows+(l-1)]=='R')&&(fl[k*rows+(l-1)]==0)){
      fl[k*rows+(l-1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l-1,cnt);
    }
    if((maz[(k+1)*rows+l]=='U')&&(fl[(k+1)*rows+l]==0)){
      fl[(k+1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k+1, l,cnt);
    }
  }
  else if((k==n)&&(l==0)){  //Bottom-Left
    //k=i;
    //l=j;
    if((maz[k*rows+(l+1)]=='L')&&(fl[k*rows+(l+1)]==0)){
      fl[k*rows+(l+1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, (l+1),cnt);
    }
    if((maz[(k-1)*rows+l]=='D')&&(fl[(k-1)*rows+l]==0)){
      fl[(k-1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k-1, l,cnt);
    }
  }
  else if((k==n)&&(l==m)){ //Bottom-Right
    //k=i;
    //l=j;
    if((maz[k*rows+(l-1)]=='R')&&(fl[k*rows+(l-1)]==0)){
      fl[k*rows+(l-1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l-1,cnt);
    }
    if((maz[(k-1)*rows+l]=='D')&&(fl[(k-1)*rows+l]==0)){
      //k--;
      fl[(k-1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k-1, l,cnt);
    }
  }
  else if(k!=0 && k!=n && l==0) { //Left side
    //k=i;
    //l=j;
    if((maz[k*rows+(l+1)]=='L')&&(fl[k*rows+(l+1)]==0)){
      fl[k*rows+(l+1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, (l+1),cnt);
    }
    if((maz[(k-1)*rows+l]=='D')&&(fl[(k-1)*rows+l]==0)){
      fl[(k-1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k-1, l,cnt);
    }
    if((maz[(k+1)*rows+l]=='U')&&(fl[(k+1)*rows+l]==0)){
      fl[(k+1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k+1, l,cnt);
    }
  }
  else if(l==m && k!=0 && k!=n) { //Rifht side
    //k=i;
    //l=j;
    if((maz[k*rows+(l-1)]=='R')&&(fl[k*rows+(l-1)]==0)){
      fl[k*rows+(l-1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l-1,cnt);
    }
    if((maz[(k-1)*rows+l]=='D')&&(fl[(k-1)*rows+l]==0)){
      fl[(k-1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k-1, l,cnt);
    }
    if((maz[(k+1)*rows+l]=='U')&&(fl[(k+1)*rows+l]==0)){
      fl[(k+1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k+1, l,cnt);
    }
  }
  else if((k==0)&&(l!=0)&&(l!=m)){ //Upper side
    //k=i;
    //l=j;
    if((maz[k*rows+(l-1)]=='R')&&(fl[k*rows+(l-1)]==0)){
      fl[k*rows+(l-1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l-1,cnt);
    }
    if((maz[k*rows+(l+1)]=='L')&&(fl[k*rows+(l+1)]==0)){
      fl[k*rows+(l+1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l+1,cnt);
    }
    if((maz[(k+1)*rows+l]=='U')&&(fl[(k+1)*rows+l]==0)){
      fl[(k+1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k+1, l,cnt);
    }
  }
  else if((k==n)&&(l!=0)&&(l!=m)){  //Bottom side
    //k=i;
    //l=j;
    if((maz[k*rows+(l-1)]=='R')&&(fl[k*rows+(l-1)]==0)){
      fl[k*rows+(l-1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l-1,cnt);
    }
    if((maz[k*rows+(l+1)]=='L')&&(fl[k*rows+(l+1)]==0)){
      fl[k*rows+(l+1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l+1,cnt);
    }
    if((maz[(k-1)*rows+l]=='D')&&(fl[(k-1)*rows+l]==0)){
      fl[(k-1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k-1, l,cnt);
    }
  }
  else {      //inside
    if((maz[k*rows+(l-1)]=='R')&&(fl[k*rows+(l-1)]==0)){
      fl[k*rows+(l-1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l-1,cnt);
    }
    if((maz[k*rows+(l+1)]=='L')&&(fl[k*rows+(l+1)]==0)){
      fl[k*rows+(l+1)]=1;
      cnt++;
      prev_room(maz,fl, n, m, k, l+1,cnt);
    }
    if((maz[(k-1)*rows+l]=='D')&&(fl[(k-1)*rows+l]==0)){
      fl[(k-1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k-1, l,cnt);
    }
    if((maz[(k+1)*rows+l]=='U')&&(fl[(k+1)*rows+l]==0)){
      fl[(k+1)*rows+l]=1;
      cnt++;
      prev_room(maz,fl, n, m, k+1, l,cnt);
    }
  }
  return cnt;
}



int main(int argc, char **argv) {
  ifstream file;
  int N,M;
  file.open(argv[1]);
  file >> N;
  file >> M;
  char maze[N*M];  //1D array with maze rooms and directions
  bool flag[N*M];  //1D flag array if this room is good
  int n1, m1,rows,res,counter=0,sum=0;
  rows=M;
  n1=N-1;
  m1=M-1;  //for recursive function
  for (int i=0;i<N;i++){
    for (int j=0;j<M;j++){
      file >> maze[i*rows+j];  //input from file
      flag[i*rows+j]=0;
    }
  }
  file.close();

  for (int i = 0; i < N; ++i) {
    for (int j = 0; j < M; ++j) {
      if ((j==0 && maze[i*rows+j]=='L') || (j == M-1 && maze[i*rows+j] == 'R') || (i == 0 && maze[i*rows+j] == 'U') || (i == N-1 && maze[i*rows+j] == 'D')) {
        flag[i*rows+j] = 1;   //mark the good rooms from the sides
        counter++;
      }
    }
  }

  for(int i=0; i<N; i++){
    for (int j=0; j<M; j++){
    if ((j==0 && maze[i*rows+j]=='L') || (j == M-1 && maze[i*rows+j] == 'R') || (i == 0 && maze[i*rows+j] == 'U') || (i == N-1 && maze[i*rows+j] == 'D')) {
        sum=prev_room(maze,flag,n1,m1,i,j,counter); //recursive function for good rooms
      }
    }
  }

  res=N*M-sum;
  cout<<res<<endl;

}
